pkg/querier/queryrange/querysharding.go

```diff
@@ -142,10 +147,15 @@ func (ast *astMapperware) Do(ctx context.Context, r queryrangebase.Request) (que
 	}
 	level.Debug(logger).Log("no-op", noop, "mapped", parsed.String())
 
+	// If the ast can't be mapped to a sharded equivalent,
+	// we can bypass the sharding engine and forward the request downstream.
```

Comment:
I’ve been thinking about this in my morning and here are my thoughts:

Inside `dynamicShardResolver.Shards`, we calculate the factor based on the stats. We cannot simply pass the stats when we create the `dynamicShardResolver` object. Imagine the following query:
```
count_over_time({app="foo"} |= "foo" [1h]) - count_over_time({app="bar"} |= "bar" [1h])
```
Here, `dynamicShardResolver.Shards` will be called twice: for `count_over_time` with `app="foo"` and for `count_over_time` with `app="bar"`.

If we got the indexStats for the whole query at `astMapperware.Do` and passed the combined stats to `shardResolverForConf` so we can use them at `dynamicShardResolver.Shards`, then, both calls to `dynamicShardResolver.Shards` would return the same `factor` which would be too big.

Passing the stats to `shardResolverForConf` may be feasible if inside `dynamicShardResolver.Shards` we could lookup the stats for only the corresponding matching groups. This would require passing to `shardResolverForConf` a map of `MatcherRange` to IndexStats.
Imo this would be overcomplicated, error-prone, and adds more complexity than the complexity we have right now.

Moreover, speaking of what we have implemented so far, we are not enforcing the limit correctly if the query is partially shardable. Imagine the following query:
```
count_over_time({app="foo"} |= "foo" [1h]) - sum_over_time({app="bar"} |= "bar" | unwrap foo [1h])
```
Here, `count_over_time({app="foo"} |= "foo" [1h])` is shardable, but `sum_over_time({app="bar"} |= "bar" | unwrap foo [1h])` is not.

Right now, we are enforcing the limit for `count_over_time({app="foo"} |= "foo" [1h])` at `dynamicShardResolver.Shards`, but we are not enforcing it on `sum_over_time({app="bar"} |= "bar" | unwrap foo [1h])` which never hits `dynamicShardResolver.Shards` as the expression is returned `unsharded` [here][1].
This is wrong as we should also get the stats for the unshardable parts of a query and enforce the limits on them.

With 9c62ef7, The limit is now enforced on both shardable and unshardable queries inside `astMapperware.Do`. Long story shorts, now `mapper.Parse` also returns the `bytesPerShard` and we use it to enforce the limit. I’ve added a new test to make sure the limits are applied correctly at `astMapperware.Do`.

[1]: https://github.com/grafana/loki/blob/6661d30889cb20eaf1b7da627b8591c6d7be71ad/pkg/logql/shardmapper.go#L284